### PR TITLE
saving gamma_d in diva

### LINKED ISSUE #438: problem with hyperparameters - TypeError
```
processing dataset for diva
processing dataset for diva
/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/site-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='sum' instead.
  warnings.warn(warning.format(ret))
/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/site-packages/torch/nn/init.py:412: UserWarning: Initializing zero-element tensors is a no-op
  warnings.warn("Initializing zero-element tensors is a no-op")

using device: cuda:0




!!!: not committed yet 
/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/utils/get_git_tag.py:17: UserWarning: !!!: not committed yet 
  warnings.warn("!!!: not committed yet")



b'44fee66'
model name: pathlist_te_sketch_diva_b44fee66_not_commited_2023md_09md_22_19_37_32_seed_0_221222
/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AUROC` will save all targets and predictions in buffer. For large datasets this may lead to large
  warnings.warn(*args, **kwargs)

 Experiment start at: 2023-09-22 19:37:33.168246
before training, model accuracy: 0.19572409987449646
[Fri Sep 22 19:37:46 2023]
Error in rule run_experiment:
    jobid: 0
    input: zoutput/benchmarks/benchmark_fbopt_pacs_full/hyperparameters.csv
    output: zoutput/benchmarks/benchmark_fbopt_pacs_full/rule_results/198.csv

RuleException:
TypeError in file /dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/exp_protocol/benchmark.smk, line 151:
bad operand type for unary -: 'NoneType'
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/exp_protocol/benchmark.smk", line 151, in __rule_run_experiment
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/exp_protocol/run_experiment.py", line 131, in run_experiment
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/compos/exp/exp_main.py", line 62, in execute
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/algos/trainers/train_fbopt.py", line 59, in before_tr
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/algos/trainers/train_fbopt.py", line 113, in eval_r_loss
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/models/model_diva.py", line 137, in cal_reg_loss
  File "/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/concurrent/futures/thread.py", line 58, in run 
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```processing dataset for diva
processing dataset for diva
/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/site-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='sum' instead.
  warnings.warn(warning.format(ret))
/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/site-packages/torch/nn/init.py:412: UserWarning: Initializing zero-element tensors is a no-op
  warnings.warn("Initializing zero-element tensors is a no-op")

using device: cuda:0




!!!: not committed yet 
/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/utils/get_git_tag.py:17: UserWarning: !!!: not committed yet 
  warnings.warn("!!!: not committed yet")



b'44fee66'
model name: pathlist_te_sketch_diva_b44fee66_not_commited_2023md_09md_22_19_37_32_seed_0_221222
/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AUROC` will save all targets and predictions in buffer. For large datasets this may lead to large
  warnings.warn(*args, **kwargs)

 Experiment start at: 2023-09-22 19:37:33.168246
before training, model accuracy: 0.19572409987449646
[Fri Sep 22 19:37:46 2023]
Error in rule run_experiment:
    jobid: 0
    input: zoutput/benchmarks/benchmark_fbopt_pacs_full/hyperparameters.csv
    output: zoutput/benchmarks/benchmark_fbopt_pacs_full/rule_results/198.csv

RuleException:
TypeError in file /dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/exp_protocol/benchmark.smk, line 151:
bad operand type for unary -: 'NoneType'
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/exp_protocol/benchmark.smk", line 151, in __rule_run_experiment
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/exp_protocol/run_experiment.py", line 131, in run_experiment
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/compos/exp/exp_main.py", line 62, in execute
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/algos/trainers/train_fbopt.py", line 59, in before_tr
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/algos/trainers/train_fbopt.py", line 113, in eval_r_loss
  File "/dss/dssmcmlfs01/pr74ze/pr74ze-dss-0001/ru48zab2/projects/domainlab2/DomainLab/domainlab/models/model_diva.py", line 137, in cal_reg_loss
  File "/dss/dsshome1/lxc0A/ru48zab2/miniconda3/envs/domainlab/lib/python3.9/concurrent/futures/thread.py", line 58, in run 
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```

### PR DESCRIPTION
**Pull request by e-dorigatti**
did not test, but gamma_d was never saved

fixes #438

### CODE REVIEW

**Review by smilesun — COMMENTED** [review]
gamma_d is saved via the decorator before constructor

**Comment by smilesun** [comment]
the problem is gamma_d is not set in benchmark

**Comment by smilesun** [comment]
but gamma_d should be used as regualrization, so it should be registered here
in hyper_init
and hyper_update

but maybe the first version we can just fix that in yaml file

**Comment by smilesun** [comment]
a better version would be to change it dynamically

File: domainlab/models/model_diva.py [hunk 1]
```diff
@@ -97,7 +99,11 @@ def hyper_init(self, functor_scheduler, trainer=None):
             """
             return functor_scheduler(
                 trainer=trainer,
-                beta_d=self.beta_d, beta_y=self.beta_y, beta_x=self.beta_x)
+                beta_d=self.beta_d,
+                beta_y=self.beta_y,
+                beta_x=self.beta_x,
+                gamma_d=self.gamma_d,
```
Comment (smilesun) [inline domainlab/models/model_diva.py:105]: do we have a way to judget whether the trainer is fbopt here?  For the moment, we could use this way, but the default DIVA model do not adjust gamma_d at all.

**Comment by e-dorigatti** [comment]
also fixes #440

File: examples/benchmark/benchmark_fbopt_pacs_full.yaml [hunk 2]
```diff
@@ -100,6 +106,7 @@ diva_fbopt:
     - msel_tr_loss
     - msel
     - es
+    - gamma_d
```
Comment (smilesun) [inline examples/benchmark/benchmark_fbopt_pacs_full.yaml:109]: depending on whether we register gamma_d to fbopt or not:  if we register, we do not need to tune gamma_d here.  Maybe we could tune the intiial value for gamma_d, since the intial value for all dynamically adjusted hyperparameters might be vital
